blocks/bricks/conv.py

```diff
@@ -177,6 +180,144 @@ def num_output_channels(self):
         return self.num_filters
 
 
+class ConvolutionalTranspose(Initializable):
+    """Performs the transpose of a 2D convolution.
+
+    Parameters
+    ----------
+    filter_size : tuple
+        The height and width of the filter (also called *kernels*).
+    num_filters : int
+        Number of filters per channel.
+    num_channels : int
+        Number of input channels in the image. For the first layer this is
+        normally 1 for grayscale images and 3 for color (RGB) images. For
+        subsequent layers this is equal to the number of filters output by
+        the previous convolutional layer. The filters are pooled over the
+        channels.
+    original_image_size : tuple
+        The height and width of the output (image or feature map).
+    batch_size : int, optional
+        Number of examples per batch. If given, this will be passed to
+        Theano convolution operator, possibly resulting in faster
+        execution.
+    image_size : tuple, optional
+        The height and width of the input (image or feature map). If given,
+        this will be passed to the Theano convolution operator, resulting
+        in possibly faster execution times.
+    step : tuple, optional
+        The step (or stride) with which to slide the filters over the
+        image. Defaults to (1, 1).
+    border_mode : {'valid', 'full'}, optional
+        The border mode to use, see :func:`scipy.signal.convolve2d` for
+        details. Defaults to 'valid'.
+    tied_biases : bool
+        If ``True``, it indicates that the biases of every filter in this
+        layer should be shared amongst all applications of that filter.
+        Setting this to ``False`` will untie the biases, yielding a
+        separate bias for every location at which the filter is applied.
+        Defaults to ``False``.
+
+    """
+    @lazy(allocation=['filter_size', 'num_filters', 'num_channels',
+                      'original_image_size'])
+    def __init__(self, filter_size, num_filters, num_channels,
+                 original_image_size, batch_size=None, image_size=(None, None),
+                 step=(1, 1), border_mode='valid', tied_biases=False,
+                 **kwargs):
+        super(ConvolutionalTranspose, self).__init__(**kwargs)
+
+        self.filter_size = filter_size
+        self.num_filters = num_filters
+        self.batch_size = batch_size
+        self.num_channels = num_channels
+        self.image_size = image_size
+        self.original_image_size = original_image_size
```

Comment:
So, maybe I'm too late, but the original size is ambiguous iff the stride is greater than 1.
However, currently the Theano Ops for `grad_wrt_inputs` require it as an input all the time, even when the stride is 1 and it could be inferred from the shapes of "topgrad" and "weights". Ideally, this would be improved in Theano.
Note that this computation when the stride is 1 would be similar to the shape inference for the forward convolution, but not exactly the same, since the padding would have to be handled differently.